### PR TITLE
OpenGL texture mipmapping

### PR DESCRIPTION
**Pull request by jonfryd**
This PR adds high quality downscaling (by OpenGL mipmapping) for picture browsing / slideshow mode via a new UI option.

Tested on my Macs (2010 iMac and 2009 Mini). Mipmapping really makes a huge difference in terms of reducing visible aliasing for down-sampled high frequencies (like e.g. hair) when browsing pictures. I noticed users on the forums complaining about "pixelated pictures", so this PR should make them happy.

Cropped screen capture from my iMac in Kodi (no mipmapping):
Image Alt

Mipmapping enabled:
Image Alt

4x zoom of small area (no mipmapping):
Image Alt

Mipmapping enabled:
Image Alt

### CODE REVIEW

**Comment by fritsch** [comment]
Thanks very much. 

@FernetMenta @popcornmix ping

**Comment by FernetMenta** [comment]
looks ok but I doubt that a considerable amount of users will notice this setting and activate it. 

@jonfryd 33% refers to GPU mem?

**Comment by jonfryd** [comment]
Exactly, @FernetMenta.

**Comment by fritsch** [comment]
We should talk about:
- Can we enable it by default - that's why I pinged you, too.

@jonfryd Could you add two pictures here so that we can see the difference?

**Comment by MartijnKaijser** [comment]
If disabled by default no one will find/us it is my guess. Could just enable it by default start of v17 and if complaints come add the advancedsetting back.

**Comment by jonfryd** [comment]
Sure, I'll add them tonight.

**Comment by fritsch** [comment]
@MartijnKaijser we need to check with Raspberry Pi - which is I pinged @popcornmix - but if needed we can also disable it for the PI only.

**Comment by fritsch** [comment]
@jonfryd no hurry - thx for bringing it up

**Comment by fritsch** [comment]
On a second look - this should not touch GLES platforms at all (right?) - if there is something we can improve for those - then we could bring that in together.

**Comment by FernetMenta** [comment]
@jonfryd do you have more info about your measurements? 33% increase of x. What is x? The lower it is, the less it matters.

@fritsch GLES is affected too

**Comment by wsnipex** [comment]
I'd assume x = GL texture size.

**Comment by FernetMenta** [comment]
@wsnipex a single texture is not of interest here but the sum of all textures loaded on the GPU.

**Comment by popcornmix** [comment]
I'll look into what effect this has on a Pi.

A few comments I can answer before then. The increase in GPU memory is +50% on all textures. We use the same mipmap layout as shown here which is 1.5x the size:
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0555c/CHDFJHAB.html

This affects all loaded textures. Worst case you may have a wall view of fanart images of 1080p. These will be 8MB each (of GPU memory). Lets say we have 5x5 posters on wall. The skin also keeps one row above and one row below loaded to improve scroll time. It also keeps any texture accessed in last 5 seconds which can be quite a lot when scrolling. There will also be a handful or skin backgrounds and posters/actors. That would be a few hundred MB of GPU memory, which will increase by 50% if mipmaps are enabled. I think this can cause problems on all Pi devices (especially Pi1) and probably most arm platforms.

I believe the correct term for this is trilinear filtering. It is not bicubical (which would be something higher quality).

There will be a significant performance cost for each texture uploaded. The GPU needs to resize the texture multiple times to produce the mipmaps. On Pi this is done by VPU code (e.g. SIMD but not hardware). For the case of scrolling a list of posters where there are several new textures a second, I suspect this extra work will be significant and result in laggy updates.

There will be a minor performance cost for each frame rendered. You need to fetch from two textures rather than one, and interpolate. I suspect this won't be very significant (Kodi is relatively simple in terms of number of triangles compared to say, a 3D game).

So my initial viewpoint is this needs to be disabled by default and optional. It will break some working setups if enabled by default.

**Comment by popcornmix** [comment]
Okay this doesn't build on GL ES. See https://www.khronos.org/opengles/sdk/docs/man/xhtml/glTexParameter.xml for valid parameters (GL_GENERATE_MIPMAP is not valid for GL ES).

**Comment by fritsch** [comment]
Using the GL_TEXTURE_MIN_FILTER with GL_LINEAR_MIPMAP_LINEAR instead?

**Comment by popcornmix** [comment]
GL_LINEAR_MIPMAP_LINEAR will use mipmaps if they exist but won't create them. The khronos document page suggested uploading each mipmap separately with glTexImage2D and different levels of detail.
Now there is a way of generating the mipmaps on the GPU as I've done that before, but I can't quite remember. Possibly: https://www.khronos.org/opengles/sdk/docs/man/xhtml/glGenerateMipmap.xml

**Comment by jonfryd** [comment]
@fritsch: I've added example images to the PR description.

@popcornmix: Cool info about OpenGL ES and the Pi. I have no experience with ES. Hopefully we can figure out a way to make this work. Let me know what I can do. I was concerned about the Pi which was a major reason that I decided not activating this by default, since the additional memory usage might be a problem for memory starved platforms. You are absolutely right about the correct term being trilinear filtering. I've amended the commit comment.

**Comment by popcornmix** [comment]
I think https://github.com/popcornmix/xbmc/commit/7d5c1d9e34a9ea6d6960588dc4307b5cf7c1712b makes GLES mipmaps work.
This PR + that patch are in last night's Milhouse build if anyone wants to test.

**Comment by jonfryd** [comment]
:+1: @popcornmix

**Comment by jonfryd** [comment]
Mipmapping still works nicely with your changes on Mac OS X, as well.

**Comment by fritsch** [comment]
@jonfryd So please pick @popcornmix's commit to your PR please

**Comment by jonfryd** [comment]
@fritsch: Done.

**Comment by fritsch** [comment]
Okay - we need to decide two things after this discussion:
a) How to avoid the advancedsetting - is this one appropriate for a GUI Setting, e.g. Use HQ GUI rendering?
b) Should we enable it by default and disable by default via our custom settings.xml that we already have in place for PI / PI2 / IMX?

For start we could invert the advancedsetting to be true by default and see where we get issues with user testing?

**Comment by FernetMenta** [comment]
I think we should enable it by default or we can lust leave it. But a finer granularity could make sense. Why enable this on the average skin element? Those don't have hair like shown on post 1 and won't benefit that much. Picture viewer or thumbs would benefit more.

**Comment by popcornmix** [comment]
Definitely disabled by default on Pi (and probably other low powered platforms). A GUI setting seems fine (where platform specific settings files can adjust the default).
As @FernetMenta says, this is currently a sledgehammer approach. It applies to the skin backgrounds and the UI skin controls which are typically displayed unscaled so have no benefit from creating the mipmaps. I wonder if mipmaps could conditionally be enabled when ``destsize < srcsize * factor`` with factor something like 0.8. The mipmap is unwanted if unscaled or scaled up in size.

**Comment by jonfryd** [comment]
OK, I totally agree with your points guys. My 2 cents...

1) Let's change the implementation so that mipmapping only affects picture and thumbnail browsing to lower the GPU memory footprint hereby avoiding the "sledgehammer" approach. Changes are required to the generic classes I think.
2) Add user-interface setting for toggling mipmapping (called e.g. "High quality rendering") in "System -> Pictures - Settings" (the default would be either *off* for low powered platforms, *on* for all others)

This would affect GL+GLES builds only, of course. Correct me if I'm wrong but it seems like even though CD3DTexture supports mipmapping it's actually not enabled at all, yet? "mipLevels" is seemingly 1 everywhere it's used. Anyway, someone else would have to do DirectX support if we choose to go down this path.

**Comment by da-anda** [comment]
I'm no 3D expert, but isn't mimpapping only needed if an object is displayed in different distances/sizes, using a defined amount of texture sizes/steps? It feels a bit odd that we sorta "misuse" this concept, with the downside of the additional memory overhead for the sizes we never need. Is there really no other way to get nicer scaling algorhythms applied, like via shaders? Mimpapping might make sense on zoom-animations within Kodi that stretch more than X percent, but for fixed sizes it feels odd?
So if using a special shader would also work, that would IMO be the way to work towards, as we want shader support for other things like blur, grayscale etc anyways.

Feel free to ignore my probably unqualified comment :)

**Comment by jonfryd** [comment]
Not an unqualified comment at all. :smile: Notice that zoom-animations is exactly what Kodi is doing when in slideshow mode, so mipmapping makes sense as a simple and fast anti-aliasing solution.

**Comment by MrMC** [comment]
I should try this on Apple A8/arm64, so far I am quite impressed at their cpu/gpu performance.

**Comment by fritsch** [comment]
@MrMC feel free to propose some code that might reduce resources - so that mipmapping is only used in the cases @popcornmix mentioned which would be a sane approach

@jonfryd will also be happy I think if he gets some help

**Comment by jonfryd** [comment]
Tonight I'm going to start working on the changes I proposed unless somebody protests. Hopefully I'll have something that's in a working state for GL+GLES in a few days.

Then I'm going to need some help with Windows+DX, please.

**Comment by jonfryd** [comment]
1) Global mipmapping is gone
2) It's possible to activate mipmapping (called "High quality downscaling") via the pictures settings UI instead

A much better implementation for sure.

Next step... DirectX integration. I think the DX changes are simple from what I've been able to google. It's just a matter of passing 0 (= max. mipmapping levels) instead of 1 for mipLevels doing "m_texture.Create" in TextureDX.cpp when IsMipmapped() is true. Any takers?

**Comment by ace20022** [comment]
@afedchin ^^

File: xbmc/guilib/Texture.cpp [hunk 1]
```diff
@@ -441,3 +442,13 @@ bool CBaseTexture::HasAlpha() const
 {
   return m_hasAlpha;
 }
+
+void CBaseTexture::SetMipmapping(bool mipmapping)
```
Comment (FernetMenta) [inline xbmc/guilib/Texture.cpp]: is there a valid scenario which requires setting this to false? if so, how do you do this after the texture has been loaded? in case you don't have answers, I would just drop the parameter

File: xbmc/pictures/GUIWindowSlideShow.cpp [hunk 2]
```diff
@@ -1144,6 +1144,12 @@ void CGUIWindowSlideShow::OnLoadPic(int iPic, int iSlideNumber, const std::strin
       return;
     }
     CLog::Log(LOGDEBUG, "Finished background loading slot %d, %d: %s", iPic, iSlideNumber, m_slides->Get(iSlideNumber)->GetPath().c_str());
+
+    if (CSettings::GetInstance().GetBool(CSettings::SETTING_PICTURES_HIGHQUALITYDOWNSCALING))
```
Comment (FernetMenta) [inline xbmc/pictures/GUIWindowSlideShow.cpp]: I would have expected this in this in SlideShowPicture where GL specific things a handled. Why should a window be concerened about mipmapping?

File: system/settings/settings.xml [hunk 3]
```diff
@@ -2019,6 +2019,15 @@
...
+    <category id="rendering" label="36552" help="36554">
+        <group id="1">
+          <setting id="pictures.highqualitydownscaling" type="boolean" label="36553" help="36555">
+            <level>1</level>
+            <default>false</default>
+            <control type="toggle" />
+          </setting>
+        </group>
+    </category>
   </section>
```
Comment (MartijnKaijser) [inline system/settings/settings.xml]: Absolutely unacceptable that this will be a separate category.

**Comment by jonfryd** [comment]
I'll look into implementing the missing DirectX support this weekend. It's a bit more than I thought work due to power-of-2 texture dimensions being required for shader operations like mipmapping. One possible solution is to allocate a bigger texture and upscale the texture (libswscale) before loading it to the GPU. OpenGL handles this behind the scene, I guess - not so with D3D.

**Comment by FernetMenta** [comment]
@afedchin could you help @jonfryd with DX?

**Comment by afedchin** [comment]
@jonfryd pick these commits please:
https://github.com/afedchin/xbmc/commit/f16e9bbe7acf7a9c37ce6a437240f1a8ca3e08c3
https://github.com/afedchin/xbmc/commit/d625ff8aa272f1b47636473fe21c154a1d5654b0
https://github.com/afedchin/xbmc/commit/a451ec892e96f41aeffedd843326a942ca16a0d9

**Comment by afedchin** [comment]
Results from DirectX
Mipmapping disabled:
2015-12-04 17-14-50 kodi
Mipmapping enabled:
2015-12-04 17-14-23 kodi

**Comment by fritsch** [comment]
I hope the images are the other way round, right?

**Comment by popcornmix** [comment]
Must admit the first looks better to me. Second is too blurry.
(But I suspect they are the right way round. mipmaps will reduce alias artefacts but will increase blurriness).

**Comment by fritsch** [comment]
The first one is a whole lot sharper and the mip map one looks like guassian blurred ... (if they are the right way arround).

**Comment by afedchin** [comment]
the images are in the right order. Another example:

High quality is off:
2015-12-04 18-29-58 kodi
High quality is on:
2015-12-04 18-29-31 kodi

**Comment by fritsch** [comment]
Sorry - I really don't like that. This effect we can have a lot cheaper ...

**Comment by FernetMenta** [comment]
> Sorry - I really don't like that. This effect we can have a lot cheaper ...

please elaborate

**Comment by fritsch** [comment]
The key idea here was to improve picture quality. Now we have mip mapping which costs a whole lot of gpu memory and the resulting image looks much more blurray than before. Which goal did we solve then?

**Comment by FernetMenta** [comment]
this code only affects pictures now so GPU memory is not a topic anymore. IMO this is an improvement and the expected result of the technique chosen.

**Comment by fritsch** [comment]
The above formula is generally true for every texture ... 

Let me rephrase: Is the result we now have what we wanted at the beginning? If yes - everything is good to go.

If not - which alternatives do we have to get non aliased scaled images while keeping sharpness?

**Comment by fritsch** [comment]
Mostly offtopic:
One other thing about pictures. We touch them pixel by pixel anyways when reading them, convert them,  and probably clamp them to the maximum texture size we have if they are bigger. If it's really about picture quality we should scale them that very moment in high quality.

@ace20022 has implemented an ffmpeg based picture format / decoder. In there a HQ method could be used to scale those right away.

**Comment by FernetMenta** [comment]
pictures need to be scaled after they were loaded. users may want to zoom in and out

**Comment by popcornmix** [comment]
Many platforms have a 2K limit on textures, so not much benefit from zooming into a picture that is being rendered at 1920 wide.

**Comment by MilhouseVH** [comment]
Can I raise the issue of scaling posters and fanart (etc.) in the GUI, as this seems to have been dropped entirely yet would potentially be a major beneficiary on some platforms.

Is it not possible to selectively use mimaps when scaling GUI assets, as had been suggested earlier in the comments? [This forum thread](http://forum.kodi.tv/showthread.php?tid=200401&pid=2116794#pid2116794) has some discussion about the issue.

Obviously as there may be a performance/memory overhead this may need to be optional etc.

**Comment by wsnipex** [comment]
even though mipmapping seems to blur/soften the picture, the aliasing in the originals(without mipmapping) is horrible. So this definitely is an improvement

**Comment by jonfryd** [comment]
Thanks, @afedchin!

By setting GL_TEXTURE_LOD_BIAS to some negative value (e.g. -1.0) on GL we can actually get the best of both worlds (i.e. anti-aliasing and sharpness). I'm testing it out right now. Seems very promising. Images coming soon.

**Comment by fritsch** [comment]
That sounds nice. Thanks @jonfryd for picking that up. -1 should pick the next "higher" (larger) mip map which should give us one level more details.

**Comment by jonfryd** [comment]
Yes, here are the samples (2x zoomed). The bias value is basically a tradeoff between more sharpness and less smooth filtering when animating in slideshow mode. Personally I like -0.5 as the "sweet spot" - provides for nice sharp, anti-aliased images and pretty smooth filtering during zoom animations. What do you guys think?

Thanks for pointing out the issue, @fritsch.

No mipmapping:
Image Alt

Mipmapping (-0.0 LOD bias):
Image Alt

Mipmapping (-0.5 LOD bias):
Image Alt

Mipmapping (-1.0 LOD bias):
Image Alt

**Comment by fritsch** [comment]
From the looks I'd say - 0.5 would be good to go - right? But it also depends on the texture size. Think of a texture 400x400 and a target texture size of 100x100 then the first mip mapping step would produce this 100x100 if we now choose one step higher then it would directly choose 400x400 and we would have no smoothing effect at all. I think that is what one sees a bit in the -1.0 step above.

But yeah - we all now know there are limits with that approach and we can choose the best it offers.

Other also shall decide which bias step to take.

Btw. as @MilhouseVH pointed out - posters and fanart are also in need. Therefore I liked that texture attribut (e.g. do mip mapping) a lot. Then this can be set by every higher level that actually wants it and then it is decided on the lower level.

And btw. could you do me a favor and squash the commits together?

1) Setting
2) General Code
3) GL
4) DirectX

then the code is much easier to review.

**Comment by wsnipex** [comment]
-0,5 looks good.
and yes, please include posters/banners/fanart

**Comment by FernetMenta** [comment]
I'd say that fanart and friends should be done on a separate PR. This PR should not turn into a never ending story.

**Comment by popcornmix** [comment]
GL_TEXTURE_LOD_BIAS doesn't exist in GLES so needs to be in non-GLES path.

**Comment by FernetMenta** [comment]
A single commit for this PR is fine. Seperate commits as @fritsch pointed out just break bisect or i.e. a setting without code that use the settings make no sense.

**Comment by popcornmix** [comment]
Can you cherry-pick https://github.com/popcornmix/xbmc/commit/4997e5e0eb9f552e7f9bad7c5d6c4b158ee3b3fc which enables mipmaps for Pi's gpu accelerated code path.

**Comment by fritsch** [comment]
@FernetMenta yes - but the architecture needs to done and this is best way implemented by a Texture attribute which can be honored later in the chain.

**Comment by fritsch** [comment]
> A single commit for this PR is fine. Seperate commits as @fritsch pointed out just break bisect or i.e. a setting without code that use the settings make no sense.

was meant as an intermediate state ... when there would be further changes for the different architectures ... and yeah you know best how much bikeshedding a setting will receive ... therefore I suggested that ... but as we are at "final step" already ... just squash all the commits together.

**Comment by afedchin** [comment]
please don't squash dx commits into one.

**Comment by fritsch** [comment]
@afedchin any reason you want them to go in separately? Or wip?

**Comment by afedchin** [comment]
@fritsch just not to forget why and what's changed

**Comment by fritsch** [comment]
Fernet is fully right. When one wants to bisect later on it is very important that every revision is able to be build / run. My suggestion before was:
Squash windows, squash gl, squash gles separateley and gui ... as we have a whole lot commits in there right now. Then bikeshedding can continue and at the end - squash to one commit make a lot of sense.

**Comment by jonfryd** [comment]
Thanks, @popcornmix. I've picked your commit.

I guess it must somehow be possible to implement the -0.5 LOD bias for GLES and DX, as well. Would be nice to see identical rendering on all platforms. Are you willing to look into this, @popcornmix and @afedchin, so we can finalize this PR or should I simply squash as described by @fritsch "as is"? It's your decision.

**Comment by jonfryd** [comment]
This PR has been stale since December. Any intentions of including this in 17? It needs to be dusted off and completed if so. Let me know.

**Comment by FernetMenta** [comment]
My request to squash commits is still open. A single commit for this PR is fine.

**Comment by jonfryd** [comment]
OK, synced my fork with upstream master, resolved conflicts and squashed all commits together. I built Kodi on my Mac and the feature still works. Haven't tested Windows or Pi, though.

**Comment by fritsch** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
I would consider this for v17. Any objections?
@popcornmix @afedchin could you please have a final look at this?

@MartijnKaijser ok with you? Could you do a test on Android?

**Comment by afedchin** [comment]
I have no objections, but this doesn't contains changes for DX yet. Will DX changes go in in separate PR?

**Comment by afedchin** [comment]
sorry for noise, I thought that there should be different commits for GL and DX changes.

**Comment by MartijnKaijser** [comment]
I'm fine with merging as I didn't notice any side effects on android. However the question that doesn't really seems answerred yet is if we should enable it by default and disable it on rpi/ios?

**Comment by popcornmix** [comment]
No objections. I can't see any change in image produced on a Pi except for double the GPU memory of the photo being used.
I believe this is because of the 2K max texture size is comparable to 1920 image width, so the photo is resized to pretty much the screen size before the texture is uploaded. Therefore you just see the full size texture and the mipmaps are not used.

File: addons/resource.language.en_gb/resources/strings.po [hunk 4]
```diff
@@ -18880,7 +18880,19 @@ msgctxt "#36618"
...
+#. Label of a setting, allow the user to specifiy high quality downscaling of pictures
+#: system/settings/settings.xml
+msgctxt "#36619"
+msgid "High quality downscaling"
+msgstr ""
+
+#. Description of setting "Pictures -> Slideshow -> High Quality Downscaling" with label #36619
+#: system/settings/settings.xml
+msgctxt "#36620"
+msgid "Enable high quality downscaling of pictures."
```
Comment (garbear) [inline addons/resource.language.en_gb/resources/strings.po]: Per martijn's suggestion, I would change this to:

"Enable high quality downscaling of pictures. This has a moderate performance impact."

**Comment by jonfryd** [comment]
I've added both suggestions (memory + performance) to the description.

File: addons/resource.language.en_gb/resources/strings.po (hunk 4 above)
Comment (FernetMenta) [inline addons/resource.language.en_gb/resources/strings.po]: > This has a moderate performance impact

not sure if we need to document the obvious. nothing is for free and if it was in this case, we wouldn't need a setting.

Comment (MartijnKaijser) [inline addons/resource.language.en_gb/resources/strings.po]: We have a description box a reason and not for c/p the setting text and adding ënable" in front of it. I'm fine with the made adjustments.

Comment (da-anda) [inline addons/resource.language.en_gb/resources/strings.po]: @FernetMenta remember, users are usually dumb, just like customers in your day job (at least in mine). I doubt anyone is thinking about any consequences or impact when happily enabling random settings, so you better explain everything in the settings description like you would to a little child to prevent a flood of support requests on the forum etc

Comment (FernetMenta) [inline addons/resource.language.en_gb/resources/strings.po]: > explain everything in the settings description like you would to a little child

but exactly this is not the case here. I would consider me a bad daddy, if I explained it like this to my daughter. she would like to know: "what does performance mean in this case? this is just a picture and it does not matter how long it takes to scale"

**Comment by MartijnKaijser** [comment]
please squash so we can merge this

**Comment by jonfryd** [comment]
@MartijnKaijser, done.

**Comment by MartijnKaijser** [comment]
jenkins build and merge

**Comment by da-anda** [comment]
we just got a comment about how mipmapping could be improved:
``` anholt APP [10:55 PM] 
took a look at kodi on the vc4 driver today: it seems at startup it's trying to set up a GL_R16 texture for dithering, even though GL_ARB_texture_rg isn't exposed.

[10:57]  
kodi is also triggering Mesa to overallocate its Y/U/V textures as full mipmap stacks.  You could reduce the GPU memory usage by setting the minification filter to linear (not mipmapped) before doing the TexImage instead of after.

anholt APP [11:16 PM] 
looks like the dither texture is getting used at runtime, and since the TexImage failed, it's just sampling the fallback texture (always 0,0,0,1).```
